File: python/lib/compass.py
from collections import deque


class Compass(object):
    NORTH = 'N'
    EAST = 'E'
    SOUTH = 'S'
    WEST = 'W'

    def __init__(self, heading):
        self.directions = deque([
            self.NORTH,
            self.EAST,
            self.SOUTH,
            self.WEST], maxlen=4)

        self.rotate_until(heading)

    def rotate_until(self, goal):
        while True:
            if self.directions[0] == goal:
                break

            self.rotate_right()

    def rotate_left(self):
        directions = self.directions
        directions.rotate(1)
        self.directions = directions

    def rotate_right(self):
        directions = self.directions
        directions.rotate(-1)
        self.directions = directions

    def direction(self):
        return self.directions[0]

File: elixir/lib/rover_challenge/compass.ex
defmodule RoverChallenge.Compass do
  @directions ["N", "E", "S", "W"]
  defstruct [directions: @directions]

  def init(heading) do
    rotate_until(%__MODULE__{}, heading)
  end

  def rotate_left(compass) do
    d = compass.directions
    %{compass | directions: [Enum.at(d, -1) | Enum.drop(d, -1)]}
  end

  def rotate_right(compass) do
    [h|t] = compass.directions
    %{compass | directions: t ++ [h]}
  end

  defp rotate_until(compass, heading) do
    if direction(compass) == heading do
      compass
    else
      compass
      |> rotate_right
      |> rotate_until(heading)
    end
  end

  def direction(compass), do: Enum.at(compass.directions, 0)
  def directions, do: @directions
end